File: apps/etl-clojure/src/pulso/xml/parser.clj
(ns pulso.xml.parser
  (:require [clojure.data.xml :as xml]
            [clojure.java.io :as io]
            [clojure.tools.logging :as log]
            [pulso.xml.io :as xml-io])
  (:import [java.io BufferedReader]))

(defn parse-health-data
  "Streams the Apple Health XML file and calls handler-fn for each top-level
   element. Uses lazy parsing via StAX so the full document is never in memory.

   handler-fn receives two args: the XML element and the locale string."
  [xml-file handler-fn]
  (log/info "Parsing XML file:" xml-file)
  (with-open [rdr (BufferedReader. (io/reader xml-file))]
    (let [clean-rdr (xml-io/skip-doctype-reader rdr)
          root      (xml/parse clean-rdr)
          locale    (get-in root [:attrs :locale])]
      (log/info "Root tag:" (:tag root) "locale:" locale)
      (doseq [element (:content root)
              :when (map? element)]
        (handler-fn element locale)))))

File: apps/etl-python/pulso/xml/parser.py
import logging
import xml.etree.ElementTree as ET

logger = logging.getLogger(__name__)


def parse_health_data(xml_file, handler_fn):
    """Streams the Apple Health XML file and calls handler_fn(element, locale)
    for each direct child of the root element. Uses ElementTree.iterparse so
    the full document is never held in memory: each processed child is
    dropped from the root's children list once handled."""
    logger.info("Parsing XML file: %s", xml_file)
    context = ET.iterparse(xml_file, events=("start", "end"))
    root = None
    locale = None
    depth = 0
    for event, elem in context:
        if event == "start":
            depth += 1
            if root is None:
                root = elem
                locale = elem.attrib.get("locale")
                logger.info("Root tag: %s locale: %s", root.tag, locale)
        else:
            depth -= 1
            if depth == 1:
                handler_fn(elem, locale)
                root.remove(elem)
